Various fixes for dev-libs/{sway{idle,lock,},wlroots} & version bump to sway-1.0_rc2

SpeedJack opened the pull request:
@prometheanfire @flappyports 
This merges the changes introduced by Matthew and those made in the old PR #10932 .
Changes are described in commit messages. Apparently, these are a lot. But these are mainly minor edits to make all ebuilds in the same format (reorder of `(R)DEPEND/HOMEPAGE/DESCRIPTION`; alphabetical order for use flags; etc.); Other are fixes to minor bugs (`wayland-protocols` and `scdoc` in BDEPEND; add missing `xkeyboard-config` dep; etc.); Some are choices, that may be argueable (`wl-clipboard` removed from dependency list; `swaybar & swaybg` use flags against only one flag `gdk-pixbuf`; `doc` -> `man` renaming; etc.).

Regarding bug #677522, I tried building every version (except the live version) of Sway without git and it works for me.

Matthew said that he opened a thread on gentoo-dev to ask for a new category `wayland-misc` or something like. I hope it will be approved - but if not, I would consider moving `dev-libs/{sway{lock,idle},wl-clipboard}` to `app-misc`, which is a better category.

prometheanfire commented:
* we should probably just remove the betas (and earlier) given their stability.
  * perhaps for wlroots too 
* Since on disk stuff is being edited sway-1.0_rc1 should be renamed to sway-1.0_rc1-r1
* MESON_AUTO_DEPEND=no should be specified if we are specifying which meson ourselves
* we should use meson_use when we can (my opinion)
* `~dev-libs/wlroots-0.3[elogind=,systemd=,X=]` is problematic unless upstream states they ONLY support that version it should be `>=`.  static deps make maintenance a pain.
* (minor) tray should be auto-enabled?
* {R,B,}DEPEND and first lines being newlines, that's done to aid readability.

Hard to tell via the diff, but latest version and 9999 should be the same (unless 9999 had changes not in latest version), this lets us copy 9999 to the new release when it comes for easy bumps.

prometheanfire commented:
I think swayidle needs systemd[polkit] to actually be allowed to send the lock

```bash
swayidle -w before-sleep 'swaylock'
Failed to send Inhibit signal: Permission denied
```

prometheanfire commented:
yep, remerged systemd and reloaded systemd-logind and now it works.  elogind probably needs the same.

SpeedJack commented:
>     * Since on disk stuff is being edited sway-1.0_rc1 should be renamed to sway-1.0_rc1-r1
> 
>     * we should use meson_use when we can (my opinion)
> 
>     * `~dev-libs/wlroots-0.3[elogind=,systemd=,X=]` is problematic unless upstream states they ONLY support that version it should be `>=`.  static deps make maintenance a pain.
> 
> I think swayidle needs `systemd[policykit]` to actually be allowed to send the lock

Done.

>     * {R,B,}DEPEND and first lines being newlines, that's done to aid readability.
>

With `{R,B,}DEPEND` you mean also to change the order of these vars? In *skel.ebuild* `DEPEND` comes before `RDEPEND`.

>     * MESON_AUTO_DEPEND=no should be specified if we are specifying which meson ourselves
> 

`MESON_AUTO_DEPEND` has been removed. But anyway `meson.eclass` has been updated 4 days ago to auto-depend on `>=meson-0.48.2` - so I've removed the explicit dependency.

>     * we should probably just remove the betas (and earlier) given their stability.
>       
>       * perhaps for wlroots too
> 

I've removed older versions of sway and wlroots except the last 2 versions of each: this would allow a user to fallback to the older version if something breaks in the latest.

>     * (minor) tray should be auto-enabled?
> 

Sway's tray feature now depends on systemd/elogind (see `REQUIRED_USE`). Default enabling it would force the user to either enable systemd/elogind or to explicity disable tray.

> Hard to tell via the diff, but latest version and 9999 should be the same (unless 9999 had changes not in latest version), this lets us copy 9999 to the new release when it comes for easy bumps.

The only difference between `sway-1.0_rc1-r1` and `sway-9999` is:
```diff
28c28
<       >=dev-libs/wlroots-0.3[elogind=,systemd=,X=]
---
>       ~dev-libs/wlroots-9999[elogind=,systemd=,X=]
```
While `sway-9999` actually works with `wlroots-0.3`, I think it's better to leave the dep on `wlroots-9999` - or we risk having to edit the ebuild each time the upstream devs add something new to wlroots and sway.

prometheanfire commented:
If we want to have wlroots-9999 for sway-9999 then we should probably lock it behind a if VER string like is done in other places.  the other changes sound good and I'll review it fully some time today.

SpeedJack commented:
I'll also add sway 1.0-rc2 (released 4h ago) to this PR later today.

prometheanfire commented:
ok, I've updated to that locally already (using the rc1 ebuild, deps didn't change it looks like).

SpeedJack commented:
> use `lddtree` and the like to check linking (and runtime deps).

I've checked. deps seems fine. I see nothing that can break.

prometheanfire commented on dev-libs/sway/sway-1.0_rc2.ebuild:
```diff
@@ -65,36 +72,35 @@ src_prepare() {
 	fi
-	if [[ ${PV} != 9999 ]]; then
-		emesonargs+=("-Dsway-version=${PV}")
-	fi
```
it looks like sway-version was removed from rc2 (and 9999)

SpeedJack replied:
You commented an outdate code. I've already removed it.

prometheanfire commented:
that's the only issue I see at this point

SpeedJack commented:
I've added a commit that forces the use of setuid instead of filecaps (explanation of why in commit message). I've done this with a revbump, not sure if needed.

prometheanfire commented:
It'd be better to not revbump from rc2 to rc2-r1 as it's not been pushed/published.  It'd be better to commit as normal, then rebase/reoder/squash your commits as needed.

prometheanfire commented on dev-libs/sway/sway-9999.ebuild:
```diff
@@ -94,7 +92,9 @@ src_configure() {
 	if ! use systemd && ! use elogind; then
-		fcaps_pkg_postinst
+		local root=${EROOT:-${ROOT}}
+		chown root:0 "${root%/}/usr/bin/sway" || die
```
fowners can't be used here?

prometheanfire commented on dev-libs/sway/sway-9999.ebuild:
```diff
@@ -94,7 +92,9 @@ src_configure() {
-		fcaps_pkg_postinst
+		local root=${EROOT:-${ROOT}}
+		chown root:0 "${root%/}/usr/bin/sway" || die
+		chmod 4711 "${root%/}/usr/bin/sway" || die
```
and fperms here?

https://devmanual.gentoo.org/function-reference/install-functions/

prometheanfire commented on dev-libs/sway/sway-1.0_rc2-r1.ebuild:
```diff
@@ -65,36 +70,37 @@ src_prepare() {
 	fi
-	if [[ ${PV} != 9999 ]]; then
-		emesonargs+=("-Dsway-version=${PV}")
-	fi
```
still missing from rc2

SpeedJack replied:
rc2 and 9999 must not have these 3 lines. `sway-version` has been removed on >=1.0_rc2.
These lines remains only in rc1 since this version still uses `sway-version`.
This causes rc1 ebuild to be different from rc2 and 9999, but i think it's fine: that's the *difference* between rc1 and later versions.

prometheanfire replied:
ah, ok, ya, that's good then.

prometheanfire commented on dev-libs/sway/sway-9999.ebuild:
```diff
@@ -65,36 +70,37 @@ src_prepare() {
 	fi
-	if [[ ${PV} != 9999 ]]; then
-		emesonargs+=("-Dsway-version=${PV}")
-	fi
```
and here

SpeedJack commented:
>  is there an upstream bug for not dropping cap_sys_admin? that's probably something upstream cares about

Mmh... well I found out that, even if they do not drop capabilities anymore, they still list the use of cap_sys_admin as a choice.
On second thought, maybe we could let the user decide whether to use cap_sys_admin or setuid (there's the `filecaps` flag for this). So I'm more inclined to just remove the commit.

prometheanfire commented:
Removing it is fine, but I'm still interested in hearing how upstream intends this to be used.

SpeedJack commented:
> I'm still interested in hearing how upstream intends this to be used.

[Sway's wiki](https://github.com/swaywm/sway/wiki/Running-Sway-without-systemd#direct-via-capabilities)
I've also added the `wlroots[filecaps=]` constraint to Sway: as per wiki, to use Sway with cap_sys_admin, wlroots must be compiled with libcap enabled.

The PR should be ok now.

prometheanfire commented:
it looks like they want to switch to setuid

```
10:29 <           emersion > ddevault, do you remmember why we don't drop caps anymore?
10:35 <           ddevault > emersion: no
10:35 <           ddevault > emersion: I just want to use setuid going forward
10:35 <           emersion > yeah, i don't think caps add any value
```

SpeedJack commented:
> it looks like they want to switch to setuid

well, we have 3 choices:
1) force the use of setuid and totally drop caps;
2) let the user decide whether to use setuid or caps, and by default use caps;
3) let the user decide whether to use setuid or caps, and by default use setuid;

Currently we are on (2), since `filecaps` is default enabled by `fcaps.eclass`.
I think we can obtain (3) by setting `-filecaps` in `IUSE`, forcing the flag to be default disabled.
Imho it's fine to leave it as is (choice 2), since it let the user to choose between caps/setuid (and I usually feel better when I can choose) and does not add a redundant flag (`-filecaps`) to the IUSE flag.

prometheanfire commented:
option 4, force setuid when !systemd and !elogind

this is a security issue

prometheanfire commented:
but for now leave it as it until upstream decides what to do

SpeedJack commented:
Upstream suggests to use only setuid: https://github.com/swaywm/sway/issues/3668
So I've removed caps in favor of setuid.
Note: done in `pkg_preinst` since `fperms` can not be used in `pkg_postinst`.

flappyports commented:
This has been merged.  We can continue to address the setuid and fcaps in future updates.
